src/scalib_ext/scalib-py/src/factor_graph.rs

```diff
@@ -141,6 +149,80 @@ fn pyobj2pubs<'a>(
     }
 }
 
+fn pyobj2genfactor_inner(py: Python, obj: &PyAny) -> PyResult<sasca::GenFactorInner> {
+    let kind: u32 = obj.getattr("kind")?.extract()?;
+    let dense: u32 = obj.getattr("GenFactorKind")?.getattr("DENSE")?.extract()?;
+    let sparse_functional: u32 = obj
+        .getattr("GenFactorKind")?
+        .getattr("SPARSE_FUNCTIONAL")?
+        .extract()?;
+    if kind == dense {
+        let factor: &numpy::PyArrayDyn<f64> = obj.getattr("factor")?.extract()?;
+        let factor = factor
+            .readonly()
+            .as_array()
+            .as_standard_layout()
+            .into_owned();
+        Ok(sasca::GenFactorInner::Dense(factor))
+    } else if kind == sparse_functional {
+        let factor: &numpy::PyArray2<sasca::ClassVal> = obj.getattr("factor")?.extract()?;
+        let factor = factor
+            .readonly()
+            .as_array()
+            .as_standard_layout()
+            .into_owned();
+        Ok(sasca::GenFactorInner::SparseFunctional(factor))
+    } else {
+        Err(PyValueError::new_err((
+            "Unknown kind",
+            obj.getattr("kind")?.to_object(py),
+        )))
+    }
+}
+
+fn pyobj2factors<'a>(
+    py: Python,
+    gen_factors: PyObject,
+    expected: impl Iterator<Item = (&'a str, bool)>,
+) -> PyResult<Vec<sasca::GenFactor>> {
+    // TODO validate single/para, dimensionality and dimensions.
```

Comment:
This TODO should be solved (the single/multi is done, but the rest is not). Either here or we need an issue for it. (It's kind of easy to mess up with that, and we should have good error messages for these cases.)

Reply:
will do this in a separate issue